### PR TITLE
Extract contract addresses from artifacts

### PR DESCRIPTION
**Pull request by nkuba**
We already download artifacts of deployed contracts from the NPM registry.
Here we add support for extracting addresses of contracts from the
artifacts. The address is extracted and stored in `_address/<CONTRACT>`
file and then embedded into the go code.

This solution is intended to be used for clients building for a specific environment. For example, if we set `CONTRACTS_NPM_PACKAGE_TAG` environment variable to `goerli`. It will download a package of contracts deployed on the Goerli network and extract addresses from it. The client will be built specifically for the Goerli network, without the need of configuring the addresses in a config file.

There will be a possibility to overwrite the addresses with a configuration file or environment variables.

Addresses extraction will happen on `go generate ./...` command execution along with the go bindings generation.

To test the solution for Goerli please run:
`CONTRACTS_NPM_PACKAGE_TAG=goerli go generate./...`

Please note that for `development` which is a default tag the addresses will be replaced with zeros, as the contracts artifacts that are published with `development` tags are intended only for ABIs generation, as the contracts are deployed locally in CI, and won't be accessible anywhere.

In a follow-up PR we will resolve contract addresses to initialize the chain handle.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -56,3 +56,4 @@ typechain/
 
 # Go bindings generator
 pkg/chain/*/*/gen/_contracts
+pkg/chain/*/*/gen/_address
```
Comment (pdyraga) [inline .gitignore:59]: Does it work for you? I see all `gen/_address` changes in my index.

Comment (nkuba) [reply .gitignore:59]: It does not! 😠

**Comment by nkuba** [comment]
> I am having problems testing it locally. Anything extra I should do?
> 
> ```
> ❯ CONTRACTS_NPM_PACKAGE_TAG=goerli go generate ./...
> rm -rf _contracts && mkdir _contracts
> npm pack --silent \
>                 --pack-destination=_contracts \
>                 @keep-network/random-beacon@2.0.0-goerli.0 \
>                 | xargs -I{} tar -zxf _contracts/{} -C _contracts
> rm -rf _address/*
> rm -rf abi/*
> rm -rf contract/*
> mkdir tmp && mv cmd/cmd*.go tmp
> rm -rf cmd/*
> mv tmp/* cmd && rm -rf tmp
> jq -jr .address /Users/piotr/git/keep-network/keep-core/pkg/chain/ethereum/beacon/gen/_contracts/package/artifacts/RandomBeacon.json > _address/RandomBeacon
> jq -jr .address /Users/piotr/git/keep-network/keep-core/pkg/chain/ethereum/beacon/gen/_contracts/package/artifacts/BeaconSortitionPool.json > _address/BeaconSortitionPool
> jq .abi /Users/piotr/git/keep-network/keep-core/pkg/chain/ethereum/beacon/gen/_contracts/package/artifacts/RandomBeacon.json > abi/RandomBeacon.abi
> go run github.com/ethereum/go-ethereum/cmd/abigen --abi abi/RandomBeacon.abi --pkg abi --type RandomBeacon --out abi/RandomBeacon.go
> go run github.com/keep-network/keep-common/tools/generators/ethlike abi/RandomBeacon.abi contract/RandomBeacon.go cmd/RandomBeacon.go
> jq .abi /Users/piotr/git/keep-network/keep-core/pkg/chain/ethereum/beacon/gen/_contracts/package/artifacts/BeaconSortitionPool.json > abi/BeaconSortitionPool.abi
> go run github.com/ethereum/go-ethereum/cmd/abigen --abi abi/BeaconSortitionPool.abi --pkg abi --type BeaconSortitionPool --out abi/BeaconSortitionPool.go
> go run github.com/keep-network/keep-common/tools/generators/ethlike abi/BeaconSortitionPool.abi contract/BeaconSortitionPool.go cmd/BeaconSortitionPool.go
> rm -rf _contracts && mkdir _contracts
> npm pack --silent \
>                 --pack-destination=_contracts \
>                  \
>                 | xargs -I{} tar -zxf _contracts/{} -C _contracts
> rm -rf _address/*
> rm -rf abi/*
> rm -rf contract/*
> mkdir tmp && mv cmd/cmd*.go tmp
> rm -rf cmd/*
> mv tmp/* cmd && rm -rf tmp
> make: *** No rule to make target `_address/WalletRegistry', needed by `all'.  Stop.
> pkg/chain/ethereum/ecdsa/gen/gen.go:8: running "make": exit status 2
> rm -rf _contracts && mkdir _contracts
> npm pack --silent \
>                 --pack-destination=_contracts \
>                  \
>                 | xargs -I{} tar -zxf _contracts/{} -C _contracts
> rm -rf _address/*
> rm -rf abi/*
> rm -rf contract/*
> mkdir tmp && mv cmd/cmd*.go tmp
> rm -rf cmd/*
> mv tmp/* cmd && rm -rf tmp
> make: *** No rule to make target `_address/Bridge', needed by `all'.  Stop.
> pkg/chain/ethereum/tbtc/gen/gen.go:8: running "make": exit status 2
> generate: open /Users/piotr/git/keep-network/keep-core/pkg/chain/ethereum/tbtc/gen/abi/Bridge.go: no such file or directory
> ```

For goerli only threshold network and random beacon generation will work. We need to publish contracts first. I suggest in this PR we test just the `development` network.

**Comment by pdyraga** [comment]
> For goerli only threshold network and random beacon generation will work. We need to publish contracts first. I suggest in this PR we test just the `development` network.

OK, this is fine. I tested with just `go generate ./...` and can confirm address extraction for development package is skipped. Also, when running for Goerli, the extraction was able to extract staking contract and random beacon contract addresses. Let's fix the problem with `.gitignore` and I am good to merge.

File: .gitignore (hunk 1 above)
Comment (pdyraga) [reply .gitignore:59]: So we can leave the `.gitignore` as is, and all we need to do is to `rm pkg/chain/ethereum/threshold/gen/_address/TokenStaking` and the rest of the files holding the address and commit this change.

These files will be generated during `go generate ./...` but not included in the git diff.

File: pkg/chain/ethereum/beacon/gen/gen.go [hunk 2]
```diff
@@ -1,4 +1,11 @@
...
 
+import (
+	_ "embed"
+)
+
 //go:generate make download_artifacts
 //go:generate make
+
+//go:embed _address/RandomBeacon
+var RandomBeaconAddress string
```
[the remark this answers is not in the document]
Comment (nkuba) [reply pkg/chain/ethereum/beacon/gen/gen.go:11]: 209bd2138cb70236452433a310c08c9bd3a61ff9

[the remark this answers is not in the document]
Comment (nkuba) [reply pkg/chain/ethereum/beacon/gen/gen.go:11]: The `#nosec` did not work.
```
Results:

Golang errors in file: [pkg/chain/ethereum/beacon/gen/gen.go]:

  > [line 11 : column 12] - pattern _address/RandomBeacon: no matching files found



Summary:
  Gosec  : 2.12.0
  Files  : 145
  Lines  : 111779
  Nosec  : 5
  Issues : 0
```

I excluded the dirs in [8abd488](https://github.com/keep-network/keep-core/pull/3086/commits/8abd488a746344d48121ea0ed63ff61f2dc5175e)

**Review by pdyraga — APPROVED** [review]

**Pull request merged**